### PR TITLE
Feat/4180  add different types of search

### PR DESCRIPTION
**Pull request by asf-algoo**
#4180 

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

**Review by raphj — APPROVED** [review]

File: frontend/src/container/AdvancedSearch.jsx [hunk 1]
```diff
@@ -333,21 +282,9 @@ export class AdvancedSearch extends React.Component {
...
-      if (searchObject.searchType === ADVANCED_SEARCH_TYPE.USER) {
-        currentSearchLength = props.userSearch.resultList.length
-      }
-
-      if (searchObject.searchType === ADVANCED_SEARCH_TYPE.SPACE) {
-        currentSearchLength = props.spaceSearch.resultList.length
-      }
-    */
-    this.getSearchResult(searchObject, currentSearchLength)
+    this.setState({ searchType: searchObject.searchType })
```
Comment (raphj) [inline frontend/src/container/AdvancedSearch.jsx:285]: Maybe check here if searchType is in Object.values(ADVANCED_SEARCH_TYPE). If not, you could fallback to contents, and maybe push a new "correct" location.

Comment (gbonaspetti) [reply frontend/src/container/AdvancedSearch.jsx]: fix

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 2]
```diff
@@ -681,10 +682,8 @@ def search_user(
                 "range", newest_authored_content_date=newest_authored_content_date_range
             )
 
-        if size:
-            search = search.extra(size=size)
-        if page_nb:
-            search = search.extra(from_=self.offset_from_pagination(size, page_nb))
+        if size and page_nb:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:685]: I don't understood this code, you should keep possibility to set size (and not just offset) and allow 0 size too (for facet only case)

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 3]
```diff
@@ -739,10 +750,8 @@ def search_workspace(
             search = search.filter("terms", member_ids=member_ids)
         if not show_deleted:
             search = search.exclude("term", is_deleted=True)
-        if size:
-            search = search.extra(size=size)
-        if page_nb:
-            search = search.extra(from_=self.offset_from_pagination(size, page_nb))
+        if size and page_nb:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:753]: Same here, i don't understood this change. It make more sense to have similar code as for content here.

**Review by inkhey — CHANGES_REQUESTED** [review]

File: backend/tracim_backend/views/search_api/schemas.py [hunk 4]
```diff
@@ -322,6 +360,19 @@ class WorkspaceSearchQuerySchema(marshmallow.Schema):
...
+        required=False, default=10, validate=strictly_positive_int_validator
+    )
+    page_nb = marshmallow.fields.Int(
+        required=False, default=1, validate=strictly_positive_int_validator
+    )
+    show_deleted = marshmallow.fields.Int(
+        example=0,
+        default=0,
+        description="if set to 1, then show deleted contents."
+        " Default is 0 - hide deleted content",
```
Comment (inkhey) [inline backend/tracim_backend/views/search_api/schemas.py:373]:
```suggestion
        description="if set to 1, then show deleted workpaces."
        " Default is 0 - hide deleted workspaces",
```
?

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 2 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:685]: ? setting `size=0/None` had no effect, even with the previous code. So this means it wasn't possible to only get facets and not results. By the way those two values (`page_nb` and `size`) are enforced to be strictly positive by the schema and defaults to `1`/`10`. So I'm even wondering whether we should care about false values here…

Comment (inkhey) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:685]: According to content search specification (last version), we should now allow 0 for size, as this permit (that's a elasticsearch feature) to have facet only search (this work with content search so i see no reason this should not work here). Also, as far i understood your fix, you remove the ability to set size, which is not something we want.

see `test_api___elasticsearch_search_ok__wildcard__facet_only` for example.

Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py]: Corrected, I wasn't aware of the `size=0` case.

**Review by inkhey — APPROVED** [review]
Backend seems ok, don't know for the frontend part.

**Review by raphj — APPROVED** [review]

**Comment by raphj** [comment]
looks good to me for everything.

**Pull request merged**